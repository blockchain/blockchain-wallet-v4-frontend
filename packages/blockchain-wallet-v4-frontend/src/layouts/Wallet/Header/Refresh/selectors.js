import { Remote } from 'blockchain-wallet-v4/src'
import { selectors } from 'data'
import { join, lift } from 'ramda'

export const getData = (state) => {
  const bitcoinContext = Remote.of(join('|', selectors.core.wallet.getWalletContext(state)))
  const etherContext = selectors.core.kvStore.ethereum.getContext(state)
  const bchContext = Remote.of(join('|', selectors.core.wallet.getWalletContext(state)))
  const path = state.router.location.pathname
  const transform = lift((bitcoinContext, etherContext, bchContext) => ({bitcoinContext, etherContext, bchContext, path}))
  return transform(bitcoinContext, etherContext, bchContext)
}
