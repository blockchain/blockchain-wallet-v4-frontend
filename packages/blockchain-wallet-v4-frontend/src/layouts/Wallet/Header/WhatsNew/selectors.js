import { selectors } from 'data'

export const getData = selectors.core.kvStore.whatsNew.getLastViewed
