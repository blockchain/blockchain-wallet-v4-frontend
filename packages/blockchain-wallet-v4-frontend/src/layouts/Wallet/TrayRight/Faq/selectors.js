import { formValueSelector } from 'redux-form'

export const getData = state => ({
  search: formValueSelector('faq')(state, 'search')
})
