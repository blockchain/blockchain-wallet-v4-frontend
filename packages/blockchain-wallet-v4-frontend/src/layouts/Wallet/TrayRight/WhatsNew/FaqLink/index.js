import React from 'react'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import { FormattedMessage } from 'react-intl'

import { Link } from 'blockchain-info-components'
import { actions } from 'data'

const FaqLink = props => (
  <Link size='13px' weight={300} onClick={() => props.actions.layoutWalletFaqClicked()}>
    <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.faqlink.learnmore' defaultMessage='Learn More' />
  </Link>
)

const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators(actions.components.layoutWallet, dispatch)
})

export default connect(undefined, mapDispatchToProps)(FaqLink)
