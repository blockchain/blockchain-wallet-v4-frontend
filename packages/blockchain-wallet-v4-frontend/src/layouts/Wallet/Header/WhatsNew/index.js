import React from 'react'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'

import { actions } from 'data'
import { getData } from './selectors'
import Error from './template.error'
import Loading from './template.loading'
import Success from './template.success'

class WhatsNewContainer extends React.Component {
  componentWillMount () {
    // this.props.actions.fetchMetadataWhatsnew()
  }

  render () {
    const { data } = this.props

    return data.cata({
      Success: (value) => <Success lastViewed={value} />,
      Failure: (message) => <Error>{message}</Error>,
      NotAsked: () => <Loading />,
      Loading: () => <Loading />
    })
  }
}

const mapStateToProps = state => ({
  data: getData(state)
})

const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators(actions.core.kvStore.whatsNew, dispatch)
})

export default connect(mapStateToProps, mapDispatchToProps)(WhatsNewContainer)
