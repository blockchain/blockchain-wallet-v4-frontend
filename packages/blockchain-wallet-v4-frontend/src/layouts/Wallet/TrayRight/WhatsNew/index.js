import React from 'react'
import styled from 'styled-components'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import { actions } from 'data'
import { getData } from './selectors'
import EmptyContent from 'services/WhatsNewService/WhatsNewContent/EmptyContent'
import { prop } from 'ramda'

export const Wrapper = styled.div`
  width: 100%;
  height: 100%;
  padding: 30px;
  box-sizing: border-box;
  overflow-y: auto;
`
const Container = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: center;
  width: 100%;
  height: auto;
`

class WhatsNewContainer extends React.PureComponent {
  componentWillUnmount () {
    this.props.kvStoreWhatsNewActions.updateMetadataWhatsNew(Date.now())
  }
  render () {
    const { announcements } = this.props
    return (
      <Wrapper>
        <Container>
          {!prop('length', announcements) ? (
            <EmptyContent />
          ) : (
            announcements.map(prop('content'))
          )}
        </Container>
      </Wrapper>
    )
  }
}

const mapDispatchToProps = dispatch => ({
  kvStoreWhatsNewActions: bindActionCreators(
    actions.core.kvStore.whatsNew,
    dispatch
  )
})

export default connect(
  getData,
  mapDispatchToProps
)(WhatsNewContainer)
