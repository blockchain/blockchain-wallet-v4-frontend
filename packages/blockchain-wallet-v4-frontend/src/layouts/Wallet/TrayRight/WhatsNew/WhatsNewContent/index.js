import React from 'react'
import { FormattedMessage } from 'react-intl'
import { LinkContainer } from 'react-router-bootstrap'

import { Link } from 'blockchain-info-components'
import FaqLink from '../FaqLink'

const LearnMore = () => <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.learnmore' defaultMessage='Learn More' />

const Announcements = [
  {
    title: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.title12' defaultMessage='Import Funds from MyEtherWallet' />,
    desc: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.desc12' defaultMessage='You can now import your ether from your MyEtherWallet backup directly into your Blockchain wallet.' />,
    date: 'February 10, 2018',
    link: <LinkContainer to='/settings/info'><Link size='13px' weight={300}><LearnMore /></Link></LinkContainer>

  }, {
    title: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.title11' defaultMessage='New Bitcoin Cash Address Format' />,
    desc: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.desc11' defaultMessage='Bitcoin Cash addresses now follow the CashAddr format.' />,
    date: 'January 18, 2018',
    link: <FaqLink />
  }, {
    title: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.title10' defaultMessage='Bitcoin Cash' />,
    desc: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.desc10' defaultMessage='If you had bitcoin in your wallet before August 1, 2017, you can now access Bitcoin Cash via Settings > General.' />,
    date: 'October 3, 2017',
    link: <LinkContainer to='/bch/transactions'><Link size='13px' weight={300}><LearnMore /></Link></LinkContainer>
  }, {
    title: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.title9' defaultMessage='Exchange Assets' />,
    desc: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.desc9' defaultMessage='You can exchange between bitcoin, ether, and bitcoin cash directly from your Blockchain wallet' />,
    date: 'August 11, 2017',
    link: <LinkContainer to='/exchange'><Link size='13px' weight={300}><LearnMore /></Link></LinkContainer>
  }, {
    title: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.title8' defaultMessage='Send & Receive Ether' />,
    desc: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.desc8' defaultMessage='You can now send and receive ether from your Blockchain wallet!' />,
    date: 'July 24, 2017',
    link: <LinkContainer to='/eth/transactions'><Link size='13px' weight={300}><LearnMore /></Link></LinkContainer>
  }, {
    title: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.title7' defaultMessage='Sell Bitcoin' />,
    desc: <FormattedMessage id='layouts.wallet.header.whatsnew.whatsnew.desc7' defaultMessage='You can now sell bitcoin directly from your Blockchain wallet!' />,
    date: 'May 12, 2017',
    link: <LinkContainer to='/buy-sell'><Link size='13px' weight={300}><LearnMore /></Link></LinkContainer>
  }
]

export default Announcements
