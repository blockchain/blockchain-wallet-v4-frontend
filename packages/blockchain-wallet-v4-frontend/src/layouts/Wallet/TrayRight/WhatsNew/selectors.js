import { filterAnnouncements } from 'services/WhatsNewService/WhatsNewContent'
import { prop } from 'ramda'
import { selectors } from 'data'
import moment from 'moment'

export const getData = state => {
  const lastViewed = selectors.core.kvStore.whatsNew
    .getLastViewed(state)
    .getOrElse(
      moment()
        .subtract(2, 'months')
        .unix()
    )
  const userCountry = selectors.core.settings
    .getCountryCode(state)
    .getOrElse(null)
  const userKycState = selectors.modules.profile
    .getUserKYCState(state)
    .getOrElse(null)

  return {
    announcements: filterAnnouncements(
      lastViewed,
      userCountry,
      userKycState
    ).filter(prop('display'))
  }
}
