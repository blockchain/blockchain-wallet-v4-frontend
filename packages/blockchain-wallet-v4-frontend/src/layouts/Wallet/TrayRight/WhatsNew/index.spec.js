import React from 'react'
import { TestBed, getDispatchSpyReducer, createTestStore } from 'utils/testbed'
import { mount } from 'enzyme'
import { combineReducers } from 'redux'
import { Remote } from 'blockchain-wallet-v4/src'
import EmptyContent from 'services/WhatsNewService/WhatsNewContent/EmptyContent'
import layoutWalletReducer from 'data/components/layoutWallet/reducers'
import WhatsNew, { Wrapper } from './index'
import { getLastViewed } from 'blockchain-wallet-v4/src/redux/kvStore/whatsNew/selectors'
import { getCountryCode } from 'blockchain-wallet-v4/src/redux/settings/selectors'
import { getUserKYCState } from 'data/modules/profile/selectors'
import { KYC_STATES } from 'data/modules/profile/model'
const { dispatchSpy, spyReducer } = getDispatchSpyReducer()

jest.useFakeTimers()

jest.mock('blockchain-wallet-v4/src/redux/kvStore/whatsNew/selectors')
jest.mock('blockchain-wallet-v4/src/redux/settings/selectors')
jest.mock('data/modules/profile/selectors')

getUserKYCState.mockImplementation(() => Remote.of(KYC_STATES.NONE))
getCountryCode.mockImplementation(() => Remote.of('US'))

describe('WhatsNew', () => {
  beforeEach(() => {
    dispatchSpy.mockClear()
  })
  const reducers = {
    spy: spyReducer,
    components: combineReducers({
      layoutWallet: layoutWalletReducer
    })
  }
  const sagas = []
  let store
  let wrapper
  beforeEach(() => {
    store = createTestStore(reducers, sagas)
    wrapper = mount(
      <TestBed withRouter={true} store={store}>
        <WhatsNew />
      </TestBed>
    )
  })

  describe('WhatsNew empty state', () => {
    getLastViewed.mockImplementation(() => Remote.of(1600142400000))
    beforeEach(() => {
      getLastViewed.mockImplementation(() => Remote.of(1600142400000))
    })
    it('should render with the empty content template', () => {
      wrapper.unmount().mount()
      wrapper.update()
      expect(wrapper.find(Wrapper)).toHaveLength(1)
      expect(wrapper.find(EmptyContent)).toHaveLength(1)
    })
  })

  describe('WhatsNew non empty state', () => {
    getLastViewed.mockImplementation(() => Remote.of(1505448000000))
    beforeEach(() => {
      getLastViewed.mockImplementation(() => Remote.of(1505448000000))
    })
    it('should not render with the empty template', () => {
      wrapper.unmount().mount()
      wrapper.update()
      expect(wrapper.find(Wrapper)).toHaveLength(1)
      expect(wrapper.find(EmptyContent)).toHaveLength(0)
    })
  })
})
