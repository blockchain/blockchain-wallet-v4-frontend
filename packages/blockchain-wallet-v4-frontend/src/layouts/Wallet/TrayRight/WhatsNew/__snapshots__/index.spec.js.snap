// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Whats New renders correctly 1`] = `
<WhatsNew__Wrapper>
  <WhatsNew__Container>
    <WhatsNew__Item
      key="0"
    >
      <WhatsNew__RowTitle>
        <text
          size="14px"
          weight={600}
        >
          <FormattedMessage
            defaultMessage="Import Funds from MyEtherWallet"
            id="layouts.wallet.header.whatsnew.whatsnew.title12"
            values={Object {}}
          />
        </text>
      </WhatsNew__RowTitle>
      <WhatsNew__RowDate>
        <text
          color="gray-3"
          size="12px"
          weight={400}
        >
          February 10, 2018
        </text>
      </WhatsNew__RowDate>
      <WhatsNew__RowContent>
        <text-group
          inline={true}
        >
          <text
            size="12px"
            weight={300}
          >
            <FormattedMessage
              defaultMessage="You can now import your ether from your MyEtherWallet backup directly into your Blockchain wallet."
              id="layouts.wallet.header.whatsnew.whatsnew.desc12"
              values={Object {}}
            />
          </text>
          <LinkContainer
            activeClassName="active"
            exact={false}
            replace={false}
            strict={false}
            to="/settings/info"
          />
        </text-group>
      </WhatsNew__RowContent>
    </WhatsNew__Item>
    <WhatsNew__Item
      key="1"
    >
      <WhatsNew__RowTitle>
        <text
          size="14px"
          weight={600}
        >
          <FormattedMessage
            defaultMessage="New Bitcoin Cash Address Format"
            id="layouts.wallet.header.whatsnew.whatsnew.title11"
            values={Object {}}
          />
        </text>
      </WhatsNew__RowTitle>
      <WhatsNew__RowDate>
        <text
          color="gray-3"
          size="12px"
          weight={400}
        >
          January 18, 2018
        </text>
      </WhatsNew__RowDate>
      <WhatsNew__RowContent>
        <text-group
          inline={true}
        >
          <text
            size="12px"
            weight={300}
          >
            <FormattedMessage
              defaultMessage="Bitcoin Cash addresses now follow the CashAddr format."
              id="layouts.wallet.header.whatsnew.whatsnew.desc11"
              values={Object {}}
            />
          </text>
          <Connect(FaqLink) />
        </text-group>
      </WhatsNew__RowContent>
    </WhatsNew__Item>
    <WhatsNew__Item
      key="2"
    >
      <WhatsNew__RowTitle>
        <text
          size="14px"
          weight={600}
        >
          <FormattedMessage
            defaultMessage="Bitcoin Cash"
            id="layouts.wallet.header.whatsnew.whatsnew.title10"
            values={Object {}}
          />
        </text>
      </WhatsNew__RowTitle>
      <WhatsNew__RowDate>
        <text
          color="gray-3"
          size="12px"
          weight={400}
        >
          October 3, 2017
        </text>
      </WhatsNew__RowDate>
      <WhatsNew__RowContent>
        <text-group
          inline={true}
        >
          <text
            size="12px"
            weight={300}
          >
            <FormattedMessage
              defaultMessage="If you had bitcoin in your wallet before August 1, 2017, you can now access Bitcoin Cash via Settings > General."
              id="layouts.wallet.header.whatsnew.whatsnew.desc10"
              values={Object {}}
            />
          </text>
          <LinkContainer
            activeClassName="active"
            exact={false}
            replace={false}
            strict={false}
            to="/bch/transactions"
          />
        </text-group>
      </WhatsNew__RowContent>
    </WhatsNew__Item>
    <WhatsNew__Item
      key="3"
    >
      <WhatsNew__RowTitle>
        <text
          size="14px"
          weight={600}
        >
          <FormattedMessage
            defaultMessage="Exchange Assets"
            id="layouts.wallet.header.whatsnew.whatsnew.title9"
            values={Object {}}
          />
        </text>
      </WhatsNew__RowTitle>
      <WhatsNew__RowDate>
        <text
          color="gray-3"
          size="12px"
          weight={400}
        >
          August 11, 2017
        </text>
      </WhatsNew__RowDate>
      <WhatsNew__RowContent>
        <text-group
          inline={true}
        >
          <text
            size="12px"
            weight={300}
          >
            <FormattedMessage
              defaultMessage="You can exchange between bitcoin, ether, and bitcoin cash directly from your Blockchain wallet"
              id="layouts.wallet.header.whatsnew.whatsnew.desc9"
              values={Object {}}
            />
          </text>
          <LinkContainer
            activeClassName="active"
            exact={false}
            replace={false}
            strict={false}
            to="/exchange"
          />
        </text-group>
      </WhatsNew__RowContent>
    </WhatsNew__Item>
    <WhatsNew__Item
      key="4"
    >
      <WhatsNew__RowTitle>
        <text
          size="14px"
          weight={600}
        >
          <FormattedMessage
            defaultMessage="Send & Receive Ether"
            id="layouts.wallet.header.whatsnew.whatsnew.title8"
            values={Object {}}
          />
        </text>
      </WhatsNew__RowTitle>
      <WhatsNew__RowDate>
        <text
          color="gray-3"
          size="12px"
          weight={400}
        >
          July 24, 2017
        </text>
      </WhatsNew__RowDate>
      <WhatsNew__RowContent>
        <text-group
          inline={true}
        >
          <text
            size="12px"
            weight={300}
          >
            <FormattedMessage
              defaultMessage="You can now send and receive ether from your Blockchain wallet!"
              id="layouts.wallet.header.whatsnew.whatsnew.desc8"
              values={Object {}}
            />
          </text>
          <LinkContainer
            activeClassName="active"
            exact={false}
            replace={false}
            strict={false}
            to="/eth/transactions"
          />
        </text-group>
      </WhatsNew__RowContent>
    </WhatsNew__Item>
    <WhatsNew__Item
      key="5"
    >
      <WhatsNew__RowTitle>
        <text
          size="14px"
          weight={600}
        >
          <FormattedMessage
            defaultMessage="Sell Bitcoin"
            id="layouts.wallet.header.whatsnew.whatsnew.title7"
            values={Object {}}
          />
        </text>
      </WhatsNew__RowTitle>
      <WhatsNew__RowDate>
        <text
          color="gray-3"
          size="12px"
          weight={400}
        >
          May 12, 2017
        </text>
      </WhatsNew__RowDate>
      <WhatsNew__RowContent>
        <text-group
          inline={true}
        >
          <text
            size="12px"
            weight={300}
          >
            <FormattedMessage
              defaultMessage="You can now sell bitcoin directly from your Blockchain wallet!"
              id="layouts.wallet.header.whatsnew.whatsnew.desc7"
              values={Object {}}
            />
          </text>
          <LinkContainer
            activeClassName="active"
            exact={false}
            replace={false}
            strict={false}
            to="/buy-sell"
          />
        </text-group>
      </WhatsNew__RowContent>
    </WhatsNew__Item>
  </WhatsNew__Container>
</WhatsNew__Wrapper>
`;
